File: som-hackathon-starter-dotnet/starters/node/src/skill-registry.ts
/**
 * In-memory registry of skill definitions, backed by JSON files in skills/.
 * Loads all *.json files on startup.
 */

import { readdirSync, readFileSync } from "fs";
import path from "path";

export interface SkillRule {
  rule_id: string;
  name: string;
  description: string;
  type: string;
  config: Record<string, any>;
  default_severity: string;
  output_message_type: string;
  affected_fields: string[];
  detail_template?: string;
}

export interface SkillDefinition {
  id: string;
  version: string;
  name: string;
  description: string;
  skill_type: string;
  disclosure_level: string;
  migration_policy: string;
  reads: string[];
  produces: string[];
  rules: SkillRule[];
}

export class SkillRegistry {
  private skills = new Map<string, SkillDefinition>();

  constructor(skillsDir?: string) {
    const dir = skillsDir ?? this.resolveSkillsDir();
    this.loadAll(dir);
  }

  all(): SkillDefinition[] {
    return [...this.skills.values()];
  }

  get(id: string): SkillDefinition | undefined {
    return this.skills.get(id);
  }

  private loadAll(dir: string): void {
    let files: string[];
    try {
      files = readdirSync(dir).filter((f) => f.endsWith(".json"));
    } catch {
      console.warn(`Skills directory not found at ${dir} — starting empty`);
      return;
    }

    for (const file of files) {
      try {
        const raw = readFileSync(path.join(dir, file), "utf-8");
        const skill: SkillDefinition = JSON.parse(raw);
        this.skills.set(skill.id, skill);
        console.log(`Loaded skill ${skill.id}@${skill.version} from ${file}`);
      } catch (err) {
        console.error(`Failed to load skill from ${file}:`, err);
      }
    }
  }

  private resolveSkillsDir(): string {
    // Prefer skills/ relative to project root (where package.json lives)
    const candidates = [
      path.join(process.cwd(), "skills"),
      path.join(__dirname, "..", "skills"),
    ];
    for (const c of candidates) {
      try {
        readdirSync(c);
        return c;
      } catch {
        // next
      }
    }
    return candidates[0]; // fallback
  }
}

File: som-hackathon-starter-dotnet/starters/python/skill_registry.py
"""
In-memory registry of skill definitions, backed by JSON files in skills/.
Loads all *.json files on startup.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass
class SkillRule:
    rule_id: str
    name: str
    description: str
    type: str
    config: dict[str, Any]
    default_severity: str
    output_message_type: str
    affected_fields: list[str]
    detail_template: str | None = None


@dataclass
class SkillDefinition:
    id: str
    version: str
    name: str
    description: str
    skill_type: str
    disclosure_level: str
    migration_policy: str
    reads: list[str]
    produces: list[str]
    rules: list[SkillRule]

    @classmethod
    def from_dict(cls, data: dict) -> SkillDefinition:
        rules = [
            SkillRule(
                rule_id=r["rule_id"],
                name=r["name"],
                description=r["description"],
                type=r["type"],
                config=r.get("config", {}),
                default_severity=r["default_severity"],
                output_message_type=r["output_message_type"],
                affected_fields=r.get("affected_fields", []),
                detail_template=r.get("detail_template"),
            )
            for r in data.get("rules", [])
        ]
        return cls(
            id=data["id"],
            version=data["version"],
            name=data["name"],
            description=data["description"],
            skill_type=data.get("skill_type", "vendor"),
            disclosure_level=data.get("disclosure_level", "L2"),
            migration_policy=data.get("migration_policy", "gated"),
            reads=data.get("reads", []),
            produces=data.get("produces", []),
            rules=rules,
        )


class SkillRegistry:
    def __init__(self, skills_dir: str | None = None):
        self._skills: dict[str, SkillDefinition] = {}
        directory = skills_dir or self._resolve_skills_dir()
        self._load_all(directory)

    def all(self) -> list[SkillDefinition]:
        return list(self._skills.values())

    def get(self, skill_id: str) -> SkillDefinition | None:
        return self._skills.get(skill_id)

    def _load_all(self, directory: str) -> None:
        path = Path(directory)
        if not path.is_dir():
            print(f"Skills directory not found at {directory} — starting empty")
            return

        for file in sorted(path.glob("*.json")):
            try:
                data = json.loads(file.read_text())
                skill = SkillDefinition.from_dict(data)
                self._skills[skill.id] = skill
                print(f"Loaded skill {skill.id}@{skill.version} from {file.name}")
            except Exception as e:
                print(f"Failed to load skill from {file.name}: {e}")

    @staticmethod
    def _resolve_skills_dir() -> str:
        candidates = [
            os.path.join(os.getcwd(), "skills"),
            os.path.join(os.path.dirname(__file__), "skills"),
        ]
        for c in candidates:
            if os.path.isdir(c):
                return c
        return candidates[0]
